### PR TITLE
Design improvements for the backend user interface

### PR DESCRIPTION
**Pull request by androettop**
I added an enterprise-like background and a file that makes it easy to change the UI color.

### CODE REVIEW

File: web_responsive/static/src/css/variables.scss
Comment (androettop) [inline web_responsive/static/src/css/variables.scss]: In this file you can modify the UI colors and variables related to font-size.

**Review by yajo — CHANGES_REQUESTED** [review]
:heart_eyes: Gorgeous!



However I observe one problem. When doing the search, the result items have very low contrast with the background:



I think it's important for usability. Could you fix that please?

Oh and also pre-commit is yelling.  Please execute `pre-commit run -a` before pushing (also use `pre-commit install` to let pre-commit forbid your commits if some linters or formatters fail).

**Comment by androettop** [comment]
Thanks you! I'm going to improve the search results style

File: .prettierrc.yml
Comment (pedrobaeza) [inline .prettierrc.yml]: Why are you changing this?

Comment (androettop) [reply .prettierrc.yml]: Because pre-commit issued the following warnings
image

Comment (pedrobaeza) [reply .prettierrc.yml]: @Yajo do you know about this?

**Comment by pedrobaeza** [comment]
It's convenient to not hammer Travis and runbot with continuous commits that launch a new build and waste CPU cycles for other PRs. Try to minimize then, please. Also, you will need eventually to squash all the commits together before merging.

**Comment by androettop** [comment]
> It's convenient to not hammer Travis and runbot with continuous commits that launch a new build and waste CPU cycles for other PRs. Try to minimize then, please. Also, you will need eventually to squash all the commits together before merging.

Understood, I have no more changes to make.
Now i'm going to simplify all the changes in a single commit. Do I keep the change of .prettierrc.yml?

**Comment by pedrobaeza** [comment]
Wait until tomorrow to see @Yajo's answer. Right now, I have made a quick test and I see a problem: when you click on the AppSwitcher application, the focus is not put on the search bar, so if you type, you don't get the search.

File: .prettierrc.yml
Comment (yajo) [reply .prettierrc.yml]: Yes, those warnings are "correct" because the `xmlWhitespaceSensitivity` setting is only used when the XML prettier plugin is loaded, and prettier runs twice: once "normal" and once with the XML plugin loaded, so it's normal that the 1st time warns about this.
Just revert this change and ignore them.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: Usually there will be theming problems in submodules when we use `!important`. Is it possible to avoid that here? Maybe by using a more specific CSS selector...

If not possible, always put a comment explaining why `!important` is required.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: The same about `!important` above.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: You're using `ease` transition below, which is more elegant. Also, animating `visibility` is pointless because it's not a numeric value.

```suggestion
        opacity: 0;
        transition: opacity 50ms ease;
```

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: It's better to stick with `rem` units, which provide a better sense of responsiveness and accessibility. Icons should be proportional to the font size.

However, I agree these could be a little bit smaller:

```suggestion
            max-width: 6rem;
```

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: Use `rem` units here too.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: I can see this code is unused. You can remove this block.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: Better to use theme variables here:

```suggestion
                    linear-gradient(to bottom, gray('200'), gray('white'));
```

File: web_responsive/static/src/xml/apps.xml
Comment (yajo) [inline web_responsive/static/src/xml/apps.xml]: Add this class to remove above styling:

```suggestion
                    <div class="search-results ml-auto mr-auto rounded" />
```

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: You can remove this styling by adding a new class as explained in other comment below:

```suggestion
```

File: web_responsive/static/src/css/variables.scss
Comment (yajo) [inline web_responsive/static/src/css/variables.scss]: You don't need to declare these variables, as [they are already found in Odoo core](https://github.com/odoo/odoo/blob/528e10572195bda2102648b935d6d8647334fd4b/addons/web/static/src/scss/primary_variables.scss#L16-L17).

File: web_responsive/views/assets.xml
Comment (yajo) [inline web_responsive/views/assets.xml]: As indicated above, all this is unnecessary:

```suggestion
```

**Review by yajo — CHANGES_REQUESTED** [review]
Thanks for this very good improvement.

Some code comments.

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: When visibility is not animated strange behavior occurs in chrome.
3163f00a-92e3-4bb5-a313-3702ff78b6ef

But if I add the animation also to the visibility this no longer happens

captured-min

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: Isn't it better to leave it in the scss to make it easier to overwrite for example in a theme?

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: > Wait until tomorrow to see @Yajo's answer. Right now, I have made a quick test and I see a problem: when you click on the AppSwitcher application, the focus is not put on the search bar, so if you type, you don't get the search.

@pedrobaeza, the problem was occurring due to the transition to display the menu, browsers prevent focus from an element that is not yet visible. I removed the transition to avoid sacrificing functionality.

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: I removed the transition to avoid sacrificing auto focus functionality, possibly in a future pull request I will add the transition keeping the auto focus.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [reply web_responsive/static/src/css/web_responsive.scss]: Something I found out when developing https://github.com/OCA/openupgradelib/pull/148 (utility to migrate from Bootstrap 3 (odoo 11-) to 4 (odoo 12+)) is that it is much easier to understand how an element is styled when you use helper classes.

When using custom css, it's much harder to know if a selector still matches an element, and if the property it is defining is stil needed or is being used fine. You need to do those checks manually always (I have migrated customer themes too), or you might end up having a lot of dead code that increases the browser payload and decreases performance, and having to refactor every now and then (as I actually did with `web_responsive` in https://github.com/OCA/web/pull/1066. With classes, it can be automated.

And I guess that some day we will have Bootstrap 5, so... well, I recommend to use classes where possible now.

---

If Odoo were not using a CSS framework, then yes. All would be easier in plain CSS. But it uses Bootstrap, so it's better to use it wherever possible.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: I still think you could leave the opacity transition. It makes it look good.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: No need to handle the `visibility` property, but we need to animate `opacity`:

```suggestion
        transition: opacity 100ms ease;
        opacity: 1;
```

(I used 100ms because it looks better IMHO, but you can use 50ms if you prefer. I'm no designer.)

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: Again, no need to deal with this:

```suggestion
```

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: And you have to animate only the animated properties here. Notice that you're animating all, which includes the inherited `opacity` value, at a higher length that its parent (300ms vs. 100ms above). That's what produces the glitch from https://github.com/OCA/web/pull/1654#discussion_r466406100:

```suggestion
                transition: 300ms ease;
                transition-property: box-shadow, transform;
```

**Review by yajo — CHANGES_REQUESTED** [review]
Follow comments below to fix the transition glitch you commented at https://github.com/OCA/web/pull/1654#discussion_r466406100.

**Comment by yajo** [comment]
I'm seeing a problem specifically with the Discuss app: when you chose it, the drawer doesn't disappear automatically:

Peek 07-08-2020 08-57

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: It is necessary to keep the visibility property because when opacity is zero the menu is invisible but it is still there. Visibility is more similar to applying a display: none;

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: Thank you!, I missed that transition, I did not understand why it happened 😅

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: As it is necessary to leave the visibility property, it is also necessary to "animate" the visibility property since if it changes without a delay, the menu disappears instantly.

Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: You can read this stackoverflow answer that explains this very well
https://stackoverflow.com/questions/27900053/css-transition-with-visibility-not-working/37905071#37905071

**Comment by androettop** [comment]
> I'm seeing a problem specifically with the Discuss app: when you chose it, the drawer doesn't disappear automatically:
> 
> Peek 07-08-2020 08-57

@Yajo , I am seeing that this problem is from before and is not caused by my changes.
I don't know the module in depth, in fact, I'm just starting with odoo, but I think this could just as the event of pressing the enter key is being handled.

captured (1)-min

Captura de pantalla de 2020-08-07 12-52-06

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [reply web_responsive/static/src/css/web_responsive.scss]: That is a good theory, but in practice, you can see that it's not needed:

Peek 10-08-2020 08-59

The difference between `visibility: invisible` and `display: none` is that an invisible block still occupies its space, while an undisplayed block doesn't.

So, this property is unneeded because Bootstrap itself is correctly managing the `display` property, and making an undisplayed block invisible is useless because it's not going to be displayed anyway. Since `display` is not animated, it changes instantly, before `opacity` starts being animated. And that's all you really need here, so `visibility` is not needed.

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [reply web_responsive/static/src/css/web_responsive.scss]: I think you forgot to include this change.

File: web_responsive/static/src/css/web_responsive.scss
Comment (androettop) [reply web_responsive/static/src/css/web_responsive.scss]: Yes, if you do this the menu effect works, and visibility is not needed, but since the display:flex property is specified, bootstraps is not managing it.

If i only manage the opacity and no the visibility (or display) property, the fade effect works, but the menu is not disappearing. You can't use the apps ui because the invisible menu is on top of it. And animation is not possible if I use display: flex / none because the change is instantaneous and the animation does not have the effect it should.

Comment (yajo) [reply web_responsive/static/src/css/web_responsive.scss]: Oh I see now! :+1: then, thanks!

File: web_responsive/static/src/css/web_responsive.scss
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]: This seems duplicated. Just remove one of them:

```suggestion
                transition-property: box-shadow, transform;
```

**Review by yajo — CHANGES_REQUESTED** [review]
Please, just one little last change.

Also could you squash the commits?

**Comment by androettop** [comment]
> Please, just one little last change.
> 
> Also could you squash the commits?

Sure, but I'm not sure how to do it, could you tell me how?

**Comment by yajo** [comment]
Yes, here it is explained very well: https://gitbetter.substack.com/p/how-to-squash-git-commits

**Review by yajo — APPROVED** [review]

**Comment by hoangtrann** [comment]
should we make these changes available in version `12.0`?

**Comment by pedrobaeza** [comment]
If you want to do it, please do the PR.

**Pull request merged**
